Lib/warnings.py

```diff
@@ -508,6 +508,130 @@ def __exit__(self, *exc_info):
         self._module._showwarnmsg_impl = self._showwarnmsg_impl
 
 
+
+class deprecated:
+    """Indicate that a class, function or overload is deprecated.
+
+    When this decorator is applied to an object, the type checker
+    will generate a diagnostic on usage of the deprecated object.
+
+    Usage:
+
+        @deprecated("Use B instead")
+        class A:
+            pass
+
+        @deprecated("Use g instead")
+        def f():
+            pass
+
+        @overload
+        @deprecated("int support is deprecated")
+        def g(x: int) -> int: ...
+        @overload
+        def g(x: str) -> int: ...
+
+    The warning specified by ``category`` will be emitted on use
+    of deprecated objects. For functions, that happens on calls;
+    for classes, on instantiation and on creation of subclasses.
+    If the ``category`` is ``None``,
+    no warning is emitted. The ``stacklevel`` determines where the
+    warning is emitted. If it is ``1`` (the default), the warning
+    is emitted at the direct caller of the deprecated object; if it
+    is higher, it is emitted further up the stack.
+
+    The deprecation message passed to the decorator is saved in the
+    ``__deprecated__`` attribute on the decorated object.
+    If applied to an overload, the decorator
+    must be after the ``@overload`` decorator for the attribute to
+    exist on the overload as returned by ``get_overloads()``.
+
+    See PEP 702 for details.
+
+    """
+    def __init__(
+        self,
+        msg: str,
+        /,
+        *,
+        category: type[Warning] | None = DeprecationWarning,
+        stacklevel: int = 1,
+    ) -> None:
+        self.msg = msg
+        self.category = category
+        self.stacklevel = stacklevel
```

Comment:
Should we forward-port https://github.com/python/typing_extensions/commit/9de9fd613913faee5db317f827f5eec3755d8a92? (If so, the tests from that commit should also be forward-ported.) Also, should the `msg`, `category` and `stacklevel` attributes be documented somewhere?

```suggestion
        if not isinstance(msg, str):
            raise TypeError(
                f"Expected an object of type str for 'msg', not {type(msg).__name__!r}"
            )
        self.msg = msg
        self.category = category
        self.stacklevel = stacklevel
```

Reply:
Thanks, applying that locally. I don't think we should document the attributes at this stage; Zac's request is not a sanctioned use for now.